maven-core/src/main/java/org/apache/maven/graph/ProjectSelector.java

```diff
@@ -0,0 +1,160 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Utility class to extract {@link MavenProject} from the project graph during the execution phase based on optional or
+ * required selectors.
+ */
+public final class ProjectSelector
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger( ProjectSelector.class );
+
+    public Set<MavenProject> getRequiredProjectsBySelectors( MavenExecutionRequest request, List<MavenProject> projects,
+                                                             Set<String> projectSelectors )
+            throws MavenExecutionException
+    {
+        Set<MavenProject> selectedProjects = new LinkedHashSet<>();
+        File baseDirectory = getBaseDirectoryFromRequest( request );
+        for ( String selector : projectSelectors )
+        {
+            Optional<MavenProject> optSelectedProject =
+                    findOptionalProjectBySelector( projects, baseDirectory, selector );
+            if ( !optSelectedProject.isPresent() )
+            {
+                String message = "Could not find the selected project in the reactor: " + selector;
+                throw new MavenExecutionException( message, request.getPom() );
+            }
+
+            MavenProject selectedProject = optSelectedProject.get();
+
+            selectedProjects.add( selectedProject );
+            selectedProjects.addAll( getChildProjects( selectedProject, request ) );
+        }
+
+        return selectedProjects;
+    }
+
+    public Set<MavenProject> getOptionalProjectsBySelectors( MavenExecutionRequest request, List<MavenProject> projects,
+                                                             Set<String> projectSelectors )
+    {
+        Set<MavenProject> resolvedOptionalProjects = new LinkedHashSet<>();
+        Set<String> unresolvedOptionalSelectors = new HashSet<>();
+        File baseDirectory = getBaseDirectoryFromRequest( request );
+        for ( String selector : projectSelectors )
+        {
+            Optional<MavenProject> optSelectedProject =
+                    findOptionalProjectBySelector( projects, baseDirectory, selector );
+            if ( optSelectedProject.isPresent() )
+            {
+                resolvedOptionalProjects.add( optSelectedProject.get() );
+                resolvedOptionalProjects.addAll( getChildProjects( optSelectedProject.get(), request ) );
+            }
+            else
+            {
+                unresolvedOptionalSelectors.add( selector );
+            }
+        }
+
+        if ( !unresolvedOptionalSelectors.isEmpty() )
+        {
+            String message = String.format( "The requested optional projects [%s] do not exist.",
+                    String.join( ",", unresolvedOptionalSelectors ) );
+            LOGGER.info( message );
```

Comment:
I think you can safely use the to string representation of the collection and use regular SLF4J placeholders here

Reply:
Tested it locally and it worked to use the string representation of the collection w/ SLF4J. 
This does result in `[:proj1,:proj2]` - which the PR does already anyway. I'd suggesting replacing the [other](https://github.com/apache/maven/blob/master/maven-core/src/main/java/org/apache/maven/DefaultMaven.java#L626) [occurrences](https://github.com/apache/maven/blob/master/maven-core/src/main/java/org/apache/maven/DefaultMaven.java#L654) as well in `DefaultMaven` to keep it consistent. Or change the current implementation to `String.join(", ", unresolvedOptionalSelectors)`. Either is fine for me.

Reply:
I'd just use default representation